sys-boot/plymouth: meson port, add: 24.004.60

DerpDays opened the pull request:
Port the existing 9999 ebuild to use meson, and adds a new ebuild for 24.004.60.

I have masked the new version under unstable keywords since its quite a large change for something related to booting.

mm1ke commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,120 @@
+DESCRIPTION="Graphical boot animation (splash) and logger"
+HOMEPAGE="https://cgit.freedesktop.org/plymouth/"
+
+LICENSE="GPL-2"
```
probably should be `GPL-2+`, see: https://cgit.freedesktop.org/plymouth/tree/src/ply-boot-server.c

DerpDays replied:
Yep, nice catch, didn't spot that so I had kept it at what it was before.

mm1ke commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,120 @@
+	# looks like make install create /var/run/plymouth
+	# this is not needed for systemd, same should hold for openrc
+	# so remove
+	rm -rf "${D}"/var/run
```
missing `|| die`

mm1ke commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,120 @@
+	dev-libs/libxslt
+	virtual/pkgconfig
+"
+BDEPEND="${DEPEND}
```
Not sure about this, but this looks strange. Don't think all `${DEPEND}` packages are required for `${BDEPEND}`?

DerpDays replied:
yeah I don't think they are needed, I'll setup the ebuildtester to try my latest commit, if it builds I'll resolve this.

mm1ke commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,120 @@
+	)
+
+	elibc_musl? ( sys-libs/rpmatch-standalone )
+	app-text/docbook-xsl-stylesheets
+	dev-libs/libxslt
+	virtual/pkgconfig
```
i guess these should belong to `BDEPEND`, but not sure about this

mm1ke replied:
also would be nice to sort these `*DEPEND` blocks unless there is a reason to have it like this.

DerpDays replied:
they should indeed be BDEPEND i think

mm1ke commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,122 @@
+	virtual/pkgconfig
+"
+
+DEPEND="
+	dev-libs/libevdev
+	drm? ( x11-libs/libdrm )
+	elibc_musl? ( sys-libs/rpmatch-standalone )
+	freetype? ( media-libs/freetype:2 )
+	gtk? (
+		dev-libs/glib:2
+		x11-libs/cairo
+		>=x11-libs/gtk+-3.14:3[X]
+	)
+	>=media-libs/libpng-1.2.16:=
+	pango? (
+		x11-libs/cairo
+		>=x11-libs/pango-1.21
+	)
+	systemd? ( sys-apps/systemd )
+	udev? ( virtual/libudev )
+	upstart-monitoring? (
+		sys-apps/dbus
+		sys-libs/ncurses
+	)
+	virtual/libc
+	x11-libs/libxkbcommon
+	x11-misc/xkeyboard-config
+"
+
+RDEPEND="${DEPEND}
+	selinux? ( sec-policy/selinux-plymouthd )
+	>sys-kernel/dracut-0.37-r3
+	udev? ( virtual/udev )
+"
```
Unconditional dependencies and conditional dependencies should be grouped together. Sorry for not being clear here.
``` suggestion
DEPEND="
	dev-libs/libevdev
	>=media-libs/libpng-1.2.16:=
	virtual/libc
	x11-libs/libxkbcommon
	x11-misc/xkeyboard-config
	drm? ( x11-libs/libdrm )
	elibc_musl? ( sys-libs/rpmatch-standalone )
	freetype? ( media-libs/freetype:2 )
	gtk? (
		dev-libs/glib:2
		x11-libs/cairo
		>=x11-libs/gtk+-3.14:3[X]
	)
	pango? (
		x11-libs/cairo
		>=x11-libs/pango-1.21
	)
	systemd? ( sys-apps/systemd )
	udev? ( virtual/libudev )
	upstart-monitoring? (
		sys-apps/dbus
		sys-libs/ncurses
	)
"

RDEPEND="${DEPEND}
	>sys-kernel/dracut-0.37-r3
	selinux? ( sec-policy/selinux-plymouthd )
	udev? ( virtual/udev )
"
```

mm1ke commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,122 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+EAPI=8
```
nitpick: can add an empty line before `EAPI=8`.. just to be inline with `skel.ebuild`

mm1ke commented:
I've added some comments. So far it looks good, however i cannot comment on the meson port. There seems to be quite some changes to the configure flags..

DerpDays commented:
I've removed the static-libs use flag since it's not actually a build option for any of the releases (i checked >=22.02.122) and on 22.02.122 ebuild it just made make try to link libs statically, which I believe didn't end up statically linking anything (or maybe just a small amount of libs as far as i can tell) and hence I felt it was kind of deceptive if it is kept (especially since there is a bug about static-libs on bugzilla i believe).

I tried to make a static build on my own but I didn't manage to get all libs statically linked and felt it too complicated to support for something that's not an upstream option

I'd be happy to include it again if you knew how to make it link everything statically though.

other new use flags:

- doc: builds documentation
- freetype: uses freetype for encryption prompts etc
- systemd: kind of self explanatory, it's systemd integration
- upstart-monitoring: there's an explanation in the manifest taken from the build options, since it is a build option and could be a feature some users want to use i added it as a use flag

I'll commit the changes you suggested in a few minutes.

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-9999.ebuild:
```diff
@@ -98,7 +108,7 @@ src_install() {
 	# this is not needed for systemd, same should hold for openrc
 	# so remove
-	rm -rf "${D}"/var/run
+	rm -rf "${D}"/var/run || die
```
`-f` nullifies the effect of `|| die`, and should therefore probably be removed.

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-9999.ebuild:
```diff
@@ -1,92 +1,102 @@
-RDEPEND="${CDEPEND}
+
+RDEPEND="${DEPEND}
+	>sys-kernel/dracut-0.37-r3
```
Is dracut really always required in this version? If so, then the warning in postinst does not make sense any more.

DerpDays replied:
Good spot, the version requirement was there in previous ebuilds: `!<sys-kernel/dracut-0.37-r3`, however I'm **assuming** that dracut is not needed at all.
The warning was also from previous ebuilds, however, dracut should indeed probably be removed, which would then make the warning make sense.

eli-schwartz replied:
dracut is NOT a version requirement in previous ebuilds.

The ! dependency in the previous ebuilds means that *iff* you have dracut installed, and you try to install plymouth, you cannot have an *old* version of plymouth. This originates back to https://bugs.gentoo.org/512436

You could remove the ! dependency because there's no need to explicitly mark plymouth as "extremely incompatible with" version 0.37 of dracut, as that version is long gone from gentoo (the restriction dates back to 2014). But that would not be removing a "dependency", that would be removing a "restriction".

AndrewAmmerlaan commented on sys-boot/plymouth/metadata.xml:
```diff
@@ -12,6 +12,10 @@
 			directories</flag>
+		<flag name="freetype">Build with freetype support (if enabled, used for
+			encryption prompts)</flag>
+		<flag name="upstart-monitoring">Listen for messages on the Upstart
+			D-Bus interface</flag>
```
We don't package upstart, so I don't think this flag is useful.

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-22.02.122-r1.ebuild:
```diff
@@ -28,7 +28,7 @@ CDEPEND="
 	gtk? (
 		dev-libs/glib:2
-		>=x11-libs/gtk+-3.14:3
+		>=x11-libs/gtk+-3.14:3[X]
```
New runtime dependencies require a revision bump

AndrewAmmerlaan replied:
That makes it an RDEPEND. CDEPEND is just a helper to have things both in RDEPEND and DEPEND, it is not something recognized by the PMS. And if we change an RDEPEND then we have to revbump to propagate those changes to users.

Alternatively, we can also simply not touch the old ebuild and focus on the new one instead.

DerpDays replied:
Oh oops! I didn't realize that this was about 22.02.122, I'll bump the revision later, the change in the gtk USE flags was due to a bug here: https://bugs.gentoo.org/850400, would of probably been better to be in its own PR though.

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,118 @@
+
+BDEPEND="
+	app-text/docbook-xsl-stylesheets
+	>=dev-build/meson-0.62
```
meson dependencies is added by the eclass and does not have to be explicitly specified

DerpDays replied:
yeah the eclass specifies a higher minimum so it can be removed

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,118 @@
+		x11-libs/cairo
+		>=x11-libs/pango-1.21
+	)
+	systemd? ( sys-apps/systemd )
```
I doubt the `systemd-integration` build option actually requires systemd. It probably just toggles the installation of service files and such. Does it produce an error if the flag is enabled but systemd is not present?

DerpDays replied:
It does fail to build without systemd present, it is required in https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/main/meson.build, and there are also a few `ifdefs` in the code for printing stuff to systemd which are also enabled with the systemd option.

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,118 @@
+
+src_configure() {
+	local emesonargs=(
+		--localstatedir "${EPREFIX}"/var
```
This I think should be:
```
-Drunstatedir=/var
```

Currently the build produces QA warnings, and this does not look like the correct style for meson options.

```
* QA Notice: The ebuild is installing to one or more unexpected directories:
*
*   /run
*
* Please fix the ebuild to use correct FHS/Gentoo policy paths.
* QA Notice: This ebuild installs into paths that should be created at runtime.
*  To fix, simply do not install into these directories.  Instead, your package
*  should create dirs on the fly at runtime as needed via init scripts/etc...
*
*   run
*   run/plymouth
* QA Notice: One or more empty directories installed to /var:
*
*   /var/lib
*   /var/lib/plymouth
*   /var/spool
*   /var/spool/plymouth
```

DerpDays replied:
I added `--localstatedir "${EPREFIX}"/var` since it would go to `/var/lib/plymouth/plymouth` without it, and also `localstatedir` was supplied in the previous ebuild 22.02.122-r1 (not on meson)

Whilst adding `-Drunstatedir=/var` removes the notice for /run, the /var notices are there, I also believe it is not appropriate to set `-Drunstatedir=/var` since in the plymouth dracut pretrigger it creates and looks for /run for plymouth (though I haven't tested or looked into what would happen with it set to /var).

I believe there should be another solution, though im not really sure what it'd be. We could try delete the /run/plymouth directory after since it is empty when created, and also it should created by the initramfs before running - so maybe we can remove it somehow? Though i'm not sure how to remove it properly in the ebuild.

For the /var notices they were still there on the previous ebuilds, should I just `keepdir` them? However, I'm unsure on what they are being used for

AndrewAmmerlaan replied:
> I also believe it is not appropriate to set `-Drunstatedir=/var` since in the plymouth dracut pretrigger it creates and looks for /run for plymouth (though I haven't tested or looked into what would happen with it set to /var).

That is in the initramfs, there is no need to have this on the rootfs right? The current ebuild does not install anything outside of FHS and is functional. The current ebuild also has a comment about deleting `/var/run` which may be related.

> For the /var notices they were still there on the previous ebuilds, should I just `keepdir` them? However, I'm unsure on what they are being used for

`keepdir` them only if they are actually used.

DerpDays replied:
>That is in the initramfs, there is no need to have this on the rootfs right? The current ebuild does not install anything outside of FHS and is functional. The current ebuild also has a comment about deleting /var/run which may be related.

Yeah your right about the initramfs my bad :sweat_smile:, the old build system probably sets `runstatedir` as `/var/run`. Though it probably doesn't need to be changed now anyways since it can stay as `/run` but be deleted at the end of the ebuild - see below on my reasoning (so it's probably better to leave it as `/run` since `/var/run` is just symlinked to `/run` due to FHS 3.0 anyways)


Right, so I've just had a quick look at the source, and from what I can tell:

`/var/lib/plymouth` is the time directory (which stores boot-duration and shutdown-duration), it is created empty during compilation, however we can remove this directory [since it is created at runtime if it doesn't already exist](https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/src/main.c?ref_type=tags#L1432)

`/var/spool/plymouth` is used to place error logs when there's errors, I think this directory can be left using `keepdir` since there's no check to ensure it is created before writing the log (not 100% sure)? (the file is linked to `/var/spool/plymouth/boot.log` using https://linux.die.net/man/3/link - which I'm not sure if it fails if the destination directory doesn't exist): error logged in these two functions https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/src/main.c?ref_type=tags#L779, which [ends up using `link` here](https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/src/libply/ply-utils.c#L651)

And finally, `/run/plymouth` is also [created at runtime similar to `/var/lib/plymouth` so it can also be deleted](https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/src/main.c?ref_type=tags#L2125)

It's probably best to double check these though.

Also if you think that its best to set `-Drunstatedir=/var` still then just let me know :sweat_smile:!

AndrewAmmerlaan replied:
> Also if you think that its best to set `-Drunstatedir=/var` still then just let me know 😅!

I suppose it doesn't really matter then since it is symlinked. For simplicity lets then not set this and use the default as it is set by the build system.

> It's probably best to double check these though.

Thanks for digging into this. Since the directories are used, I think `keepdir` is the best solution. This avoids surprises if something (in plymouth or some other package) in the future changes and fails when this directory has (not yet) been created.

eli-schwartz replied:
> This I think should be:
> 
> ```
> -Drunstatedir=/var
> ```
> Currently the build produces QA warnings, and this does not look like the correct style for meson options.

In meson.eclass, the default meson arguments includes:

```bash
        MESONARGS+=(
                --libdir "$(get_libdir)"
                --localstatedir "${EPREFIX}/var/lib"
                --prefix "${EPREFIX}/usr"
                --sysconfdir "${EPREFIX}/etc"
```

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -0,0 +1,118 @@
+BDEPEND="
+	app-text/docbook-xsl-stylesheets
+	>=dev-build/meson-0.62
+	dev-libs/libxslt
```
and probably this one too

DerpDays replied:
Could you elaborate on what is wrong with this part?

AndrewAmmerlaan replied:
BDEPEND is for things we *execute* during build, DEPEND is for things we need, but do not execute during build. This is a library and should therefore be a DEPEND. Same holds fort the stylesheets, I doubt there is any executable code there.

DerpDays replied:
for `app-text/docbook-xsl-stylesheets` a large portion of packages in the reverse dependency search were BDEPEND, so I thought it would be better to place it there.

The same applies for `dev-libs/libxslt`, I saw that quite a lot of packages have it as a BDEPEND only, though there are quite a lot that DEPEND on `libxslt` 

I've not looked enough to really know, I'll move these later today if you think they need to be DEPEND though

AndrewAmmerlaan replied:
I suppose it might be executing one of these scripts:
```
equery f dev-libs/libxslt
* Searching for libxslt in dev-libs ...
* Contents of dev-libs/libxslt-1.1.39:
/usr
/usr/bin
/usr/bin/i686-pc-linux-gnu-xslt-config
/usr/bin/x86_64-pc-linux-gnu-xslt-config
/usr/bin/xslt-config -> x86_64-pc-linux-gnu-xslt-config
/usr/bin/xsltproc
```

If so then BDEPEND is correct, if it only uses the library instead then DEPEND is correct.

DerpDays replied:
[`libxlst` is executed at build time only for docs](https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/24.004.60/docs/meson.build?ref_type=tags&blame=0#L1) from what I can see, so it can probably go under the `doc` USE flag. As for `app-text/docbook-xsl-stylesheets` they're also used only for docs (from what I can tell) and so can also be a `BDEPEND`.

I'll put them under the `doc` USE flag but keep them in `BDEPEND`, let me know if you find that it needs changing :+1:.

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-9999.ebuild:
```diff
@@ -1,113 +1,118 @@
+	x11-libs/libxkbcommon
+	x11-misc/xkeyboard-config
 	drm? ( x11-libs/libdrm )
+	elibc_musl? ( sys-libs/rpmatch-standalone )
```
This now ends up in `RDEPEND`, I don't think that is correct.

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-9999.ebuild:
```diff
@@ -1,113 +1,118 @@
+DEPEND="
+	dev-libs/libevdev
 	>=media-libs/libpng-1.2.16:=
+	virtual/libc
```
libc is in the system set and does not have to specified as a dependency

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-9999.ebuild:
```diff
@@ -1,113 +1,118 @@
 
-src_install() {
-	default
+src_compile() {
+	meson_src_compile
+}
```
This is the default `src_compile` when `meson.eclass` is inherited. So no need to specify this manually.

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -97,14 +99,27 @@ src_install() {
+	keepdir /var/spool/plymouth
+	# /var/lib/plymouth is created at runtime, and is used to store boot/shutdown
+	# durations, it doesn't need to be created at build.
+	rm -rf "${ED}"/var/lib || die
+	# /run/plymouth is also created at runtime
+	rm -rf "${ED}"/run || die
```
The `-f` argument effectively negates `|| die` since `rm` will then always `exit 0`. For this reason `-f` should usually not be used in ebuilds.

DerpDays replied:
Oh wow, literally just noticed this as well and had put the commit before I saw this, done!

AndrewAmmerlaan commented on sys-boot/plymouth/plymouth-24.004.60.ebuild:
```diff
@@ -97,14 +99,27 @@ src_install() {
+		ewarn "if you wish to have an initramfs with plymouth support, you can"
+		ewarn "emerge 'sys-kernel/dracut' or genkernel-next[plymouth] otherwise"
+		ewarn "you can look at the plymouth wiki for other methods:"
+		ewarn "https://wiki.gentoo.org/wiki/Plymouth#Building_Initramfs"
```
I don't know much about genkernel, but my understanding is that genkernel-next has been merged into genkernel and/or discontinued, so I think this `has_version` is outdated and depending on the status of plymouth support in genkernel it should either be reworded or removed.

DerpDays replied:
done!

DerpDays replied:
set the `genkernel-next` stuff as outdated on the plymouth wiki

AndrewAmmerlaan replied:
Thanks

AndrewAmmerlaan commented:
For some reason, this version of plymouth is segfaulting on my end:
```
Aug 15 12:00:20 nowa-gentoo-pc systemd[1]: plymouth-start.service: Main process exited, code=dumped, status=11/SEGV
Aug 15 12:00:20 nowa-gentoo-pc systemd[1]: plymouth-start.service: Failed with result 'core-dump'.
```

DerpDays commented:
Oh weird, works fine on openrc for me, I don't have a systemd machine on hand to test and won't be able to make one for a few days. I'll look into it later and see if it's anything obvious, otherwise I'll try replicate it when I can

AndrewAmmerlaan commented:
> Oh weird, works fine on openrc for me, I don't have a systemd machine on hand to test and won't be able to make one for a few days. I'll look into it later and see if it's anything obvious, otherwise I'll try replicate it when I can

I wonder, is it at all possible to obtain the coredump from Plymouth? I could not find it in the usual location, presumably because it coredumps in the initramfs stage.

DerpDays commented:
Does anything get written to /var/spool/plymouth/boot.log? Otherwise there's also the details theme which I believe shows info while booting, and there are also a kernel flag you can set `plymouth.debug` that might make it show some info before it crashes.

Finally we can also try copy the core dump to the main disk after the initramfs is done using `rd.break` or some similar flag see https://www.man7.org/linux/man-pages/man7/dracut.cmdline.7.html

DerpDays commented:
Can you try show the splash inside x11 or a VT (where we can get logs easier) and see if the issue persists there, because if so it would be a lot easier to debug, also if you haven't tried already, please try doing `plymouth-set-default-theme -R your_theme` or swapping it to a default theme such as `details` so we can get the easy debug option out of the way.

Theres also some extra info on debugging here if you have extra time: https://gitlab.freedesktop.org/plymouth/plymouth/-/blob/main/docs/development.txt

This describes `plymouth.debug` kernel flag and where to find logs that might have been produced etc, along with other methods of debugging.

I'll be able to try reproduce it in a few days if you don't have time or weren't able to figure it out by then

DerpDays commented:
@AndrewAmmerlaan I've created a VM using systemd, however `plymouth-24.004.60` works fine on my end, could you tell me which USE flags your using and also any `themes` and your `/etc/plymouth/plymouthd.conf`, just anything really to help me try reproduce it?

AndrewAmmerlaan commented:
> @AndrewAmmerlaan I've created a VM using systemd, however `plymouth-24.004.60` works fine on my end, could you tell me which USE flags your using and also any `themes` and your `/etc/plymouth/plymouthd.conf`, just anything really to help me try reproduce it?

I was using `USE="drm gtk pango udev systemd"` and also tried with both `freetype` enabled and disabled. 

```
cat /etc/plymouth/plymouthd.conf
[Daemon]
Theme=bgrt
```

I suspect the issue might be caused by some upgrade in some dependency because I remember runtime testing this PR earlier and while it looked different  compared to version 22.02.122-r1 there was no segfault.

DerpDays commented:
> I was using `USE="drm gtk pango udev systemd"` and also tried with both `freetype` enabled and disabled.

Works for me here on the same use flags, so it's probably not that, so its probably like you said something to do with some dependency on your system. Are all plymouth dependencies are up to date with `emerge --deep plymouth`? I'm not too sure on what else I can easily do from my side to debug this since I'm unable to reproduce this.

immolo commented:
I'm currently testing the 24.006.40.ebuild and am seeing this error while trying to manifest:

`External commands disabled while sourcing the ebuild`

Seems related to line 632.

immolo commented:
Ignore the previous, the patch gotten eaten in the pastebin I was using yet somehow applied.

Early testing shows it working on my stable openrc system, will report back with more information later on.

immolo commented:
@AndrewAmmerlaan This has now been tested on both OpenRC and Systemd with a multitude of default and custom themes and I can report this PR works and has the side benefit of being much faster to load than the previous in tree version (at least on my test machines.)

AndrewAmmerlaan commented:
It still segfaults and fails to do anything on my end:

Here's the backtrace:
```
nowa-gentoo-pc nowa # gdb /sbin/plymouthd
GNU gdb (Gentoo 15.1 vanilla) 15.1
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /sbin/plymouthd...
(No debugging symbols found in /sbin/plymouthd)
(gdb) run --no-daemon --debug --tty=/dev/tty
Starting program: /usr/bin/plymouthd --no-daemon --debug --tty=/dev/tty
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".

Program received signal SIGPIPE, Broken pipe.
0x00007ffff7d20224 in write () from /usr/lib64/libc.so.6

(gdb)
The program being debugged has been started already.
Start it from the beginning? (y or n)
Starting program: /usr/bin/plymouthd --no-daemon --debug --tty=/dev/tty
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
[New Thread 0x7ffff4a006c0 (LWP 4950)]

Thread 1 "plymouthd" received signal SIGSEGV, Segmentation fault.
0x00007ffff7f51444 in ply_terminal_set_disabled_input () from /usr/lib64/libply-splash-core.so.5

(gdb)
#0  0x00007ffff7f51444 in ply_terminal_set_disabled_input () at /usr/lib64/libply-splash-core.so.5
#1  0x00007ffff7fb0c22 in ??? () at /usr/lib64/plymouth/renderers/drm.so
#2  0x00007ffff7f5005d in ply_renderer_open_input_source () at /usr/lib64/libply-splash-core.so.5
#3  0x00007ffff7f52ab8 in ply_keyboard_watch_for_input () at /usr/lib64/libply-splash-core.so.5
#4  0x00007ffff7f52b53 in ply_device_manager_activate_keyboards () at /usr/lib64/libply-splash-core.so.5
#5  0x000055555555c790 in ??? ()
#6  0x000055555555d3dd in ??? ()
#7  0x000055555555ddb8 in ??? ()
#8  0x000055555555fd2a in ??? ()
#9  0x00007ffff7f6c65a in ply_event_loop_process_pending_events () at /usr/lib64/libply.so.5
#10 0x00007ffff7f6c968 in ply_event_loop_run () at /usr/lib64/libply.so.5
#11 0x000055555555b1e8 in main ()
```

AndrewAmmerlaan commented:
I cannot reproduce my issue on my laptop, it seems specific to this system somehow.

AndrewAmmerlaan commented:
Here's the backtrace with debug info:

```
nowa-gentoo-pc nowa # gdb /sbin/plymouthd
GNU gdb (Gentoo 15.1 vanilla) 15.1
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /sbin/plymouthd...
(No debugging symbols found in /sbin/plymouthd)
(gdb) run --no-daemon --debug --tty=/dev/tty
Starting program: /usr/bin/plymouthd --no-daemon --debug --tty=/dev/tty
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
[New Thread 0x7ffff4a006c0 (LWP 21318)]

Thread 1 "plymouthd" received signal SIGSEGV, Segmentation fault.
ply_terminal_set_disabled_input (terminal=0x0)
at ../plymouth-24.004.60/src/libply-splash-core/ply-terminal.c:335
335             if (!terminal->is_disabled) {

(gdb)
#0  ply_terminal_set_disabled_input (terminal=0x0)
at ../plymouth-24.004.60/src/libply-splash-core/ply-terminal.c:335
#1  0x00007ffff7fb0c22 in watch_input_device (backend=0x5555555a7ad0, input_device=<optimized out>)
at ../plymouth-24.004.60/src/plugins/renderers/drm/plugin.c:1869
#2  watch_input_devices (backend=0x5555555a7ad0)
at ../plymouth-24.004.60/src/plugins/renderers/drm/plugin.c:1881
#3  open_input_source (backend=0x5555555a7ad0, input_source=0x5555555a7af8)
at ../plymouth-24.004.60/src/plugins/renderers/drm/plugin.c:1895
#4  0x00007ffff7f5005d in ply_renderer_open_input_source
(renderer=0x5555555a8f50, input_source=<optimized out>)
at ../plymouth-24.004.60/src/libply-splash-core/ply-renderer.c:414
#5  0x00007ffff7f52ab8 in ply_keyboard_watch_for_renderer_input (keyboard=0x5555555a8d30)
at ../plymouth-24.004.60/src/libply-splash-core/ply-keyboard.c:378
#6  ply_keyboard_watch_for_input (keyboard=keyboard@entry=0x5555555a8d30)
at ../plymouth-24.004.60/src/libply-splash-core/ply-keyboard.c:470
#7  0x00007ffff7f52b53 in ply_device_manager_activate_keyboards (manager=0x55555556d100)
at ../plymouth-24.004.60/src/libply-splash-core/ply-device-manager.c:1401
#8  0x000055555555c790 in ??? ()
#9  0x000055555555d3dd in ??? ()
#10 0x000055555555ddb8 in ??? ()
#11 0x000055555555fd2a in ??? ()
#12 0x00007ffff7f6c65a in ply_event_loop_handle_met_status_for_source
(loop=<optimized out>, source=0x555555574da0, status=PLY_EVENT_LOOP_FD_STATUS_HAS_DATA)
at ../plymouth-24.004.60/src/libply/ply-event-loop.c:1040
#13 ply_event_loop_process_pending_events (loop=0x5555555663e0)
at ../plymouth-24.004.60/src/libply/ply-event-loop.c:1279
#14 ply_event_loop_process_pending_events (loop=loop@entry=0x5555555663e0)
--Type <RET> for more, q to quit, c to continue without paging--
at ../plymouth-24.004.60/src/libply/ply-event-loop.c:1210
#15 0x00007ffff7f6c968 in ply_event_loop_run (loop=0x5555555663e0)
at ../plymouth-24.004.60/src/libply/ply-event-loop.c:1311
#16 0x000055555555b1e8 in main ()
```

AndrewAmmerlaan commented:
Reported as https://gitlab.freedesktop.org/plymouth/plymouth/-/issues/265

EssenSea commented:
current version plymouth in gentoo repository cannot work  with graphic theme( graphic  will flash again and again but not start sddm)（openrc+elogind, with drm gtk pango udev flags) , only "details" theme can work normally. I'm not sure whether it's relative to version reason. could I got some help? THX!

Nowa-Ammerlaan commented:
Whatever issue I am experiencing is resolved upstream because the 9999 works a lot better for me.

I have rebased and merged this PR, but masked the new 24.004.60 version so folks who run into issues with the current version can opt-in for the upgrade while avoiding breaking things for other users. For the next upstream release we can revisit this and hopefully get a new version that works for everyone into the tree.

Thanks :+1:

Nowa-Ammerlaan commented:
> Btw, please remove [X] from pango dep, it is not needed.

It fails to build without it: https://github.com/dracut-ng/dracut-ng/issues/732

ganguin commented:
> > > Btw, please remove [X] from pango dep, it is not needed.
> > 
> > 
> > It fails to build without it: [dracut-ng/dracut-ng#732](https://github.com/dracut-ng/dracut-ng/issues/732)
> 
> Interesting, works for me. Using local ebuild.
> 
> ```
>     pango? ( >=x11-libs/pango-1.21 )
> ```
> 
> ```
>     [I] sys-boot/plymouth
>      Available versions:  22.02.122-r3 [M](~)24.004.60-r1 (**)9999*l (**)9999*l[1] {debug doc +drm freetype (+)gtk +pango selinux +split-usr static-libs systemd +udev}
>      Installed versions:  9999*l[1](06:37:32 PM 10/18/2024)(drm pango systemd udev -debug -gtk -selinux -split-usr)
> ```

Same for me, compiles and works without `pango[X]` since long time. `pango? ( x11-base/xorg-proto )` is not needed as well when gtk is disabled. X use flag and x11-base/xorg-proto should be guarded by the gtk use flag.

Nowa-Ammerlaan commented:
Please open a new pull request so we can properly discuss this. The dependency may be wrong, but it's clear from the failure in the Dracut CI that somewhere we do have to depend on `x11-base/xorg-proto` and possibly `x11-libs/pango[X]` though I have indeed not tested if the `gtk` flag plays a role in this.
